crypto/fipsmodule/ec/p384.c

```diff
@@ -0,0 +1,452 @@
+/*
+------------------------------------------------------------------------------------
+ Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ SPDX-License-Identifier: Apache-2.0
+------------------------------------------------------------------------------------
+*/
+
+// Implementation of P-384 that uses Fiat-crypto for the field arithmetic
+// found in third_party/fiat, similarly to p256.c
+
+#include <openssl/ec.h>
+
+#include <openssl/bn.h>
+#include <openssl/err.h>
+#include <openssl/mem.h>
+
+#include "../bn/internal.h"
+#include "../delocate.h"
+#include "internal.h"
+
+#if defined(BORINGSSL_HAS_UINT128)
+#define BORINGSSL_NISTP384_64BIT 1
+#include "../../../third_party/fiat/p384_64.h"
+#else
+#include "../../../third_party/fiat/p384_32.h"
+#endif
+
+#if defined(BORINGSSL_NISTP384_64BIT)
+#define FIAT_P384_NLIMBS 6
+typedef uint64_t fiat_p384_limb_t;
+typedef uint64_t fiat_p384_felem[FIAT_P384_NLIMBS];
+static const fiat_p384_felem fiat_p384_one = {0xffffffff00000001, 0xffffffff,
+                                              0x1, 0x0, 0x0, 0x0};
+#else  // 64BIT; else 32BIT
+#define FIAT_P384_NLIMBS 12
+typedef uint32_t fiat_p384_limb_t;
+typedef uint32_t fiat_p384_felem[FIAT_P384_NLIMBS];
+static const fiat_p384_felem fiat_p384_one = {
+    0x1, 0xffffffff, 0xffffffff, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0};
+#endif  // 64BIT
+
+static fiat_p384_limb_t fiat_p384_nz(
+    const fiat_p384_limb_t in1[FIAT_P384_NLIMBS]) {
+  fiat_p384_limb_t ret;
+  fiat_p384_nonzero(&ret, in1);
+  return ret;
+}
+
+static void fiat_p384_copy(fiat_p384_limb_t out[FIAT_P384_NLIMBS],
+                           const fiat_p384_limb_t in1[FIAT_P384_NLIMBS]) {
+  for (size_t i = 0; i < FIAT_P384_NLIMBS; i++) {
+    out[i] = in1[i];
+  }
+}
+
+static void fiat_p384_cmovznz(fiat_p384_limb_t out[FIAT_P384_NLIMBS],
+                              fiat_p384_limb_t t,
+                              const fiat_p384_limb_t z[FIAT_P384_NLIMBS],
+                              const fiat_p384_limb_t nz[FIAT_P384_NLIMBS]) {
+  fiat_p384_selectznz(out, !!t, z, nz);
+}
+
+static void fiat_p384_from_generic(fiat_p384_felem out, const EC_FELEM *in) {
+  fiat_p384_from_bytes(out, in->bytes);
+}
+
+static void fiat_p384_to_generic(EC_FELEM *out, const fiat_p384_felem in) {
+  // This works because 384 is a multiple of 64, so there are no excess bytes to
+  // zero when rounding up to |BN_ULONG|s.
+  OPENSSL_STATIC_ASSERT(
+      384 / 8 == sizeof(BN_ULONG) * ((384 + BN_BITS2 - 1) / BN_BITS2),
+      fiat_p384_to_bytes_leaves_bytes_uninitialized);
+  fiat_p384_to_bytes(out->bytes, in);
```

Comment:
Since this is not a problem atm and I can't imagine a world in which we would ever define `EC_MAX_BYTES` to anything that can support below 512, we are good.